### ISSUE TITLE
Will your 81 Tiles revamp cover Outside Connections?

### ISSUE THREAD

**Issue by originalfoo** [Quistar-LAB/EManagersLib#4]
I'm not sure if this is relevant to your planned `81 Tiles` rewrite/implementation, but assuming you also plan to handle outside connections as part of that, which would make a lot of sense, here's some notable issues that exist with current implementations...

### Intercity buses enter via _virtual lane_ on outside highway connections

Problems can occur when outside connection is not a highway, and it breaks completely if the outside connection is 2-way traffic (even on 2-way highways).

More info: https://github.com/CitiesSkylinesMods/TMPE/issues/960

![image](https://user-images.githubusercontent.com/1386719/85475432-fdc90200-b5ad-11ea-9116-1f1b975a0c25.png)

It's out of scope for TMPE but would definitely be in-scope for any mod handling outside connections.

### Zoned roads can't be used as outside connection

User has to disable zoning on the segment used for outside connection; most users aren't aware of that, and it requires separate mod to do it.

### Roads without outside connection flag don't work

`Any Road Outside Connection` mod fixes that by adding the flag to all roads IIRC.

### Train tracks must use vanilla station track at outside connection

Vanilla tracks will automatically switch to station track at outside connection - but workshop tracks don't do that (not all workshop tracks have associated station track). Would be nice if:

1. Connecting non-station track (even if non-vanilla) to outside would auto-convert to a station track (vanilla would be fine, at least the connection would work then)
2. Ability for user to use any station track (including non-vanilla) as outside connection

### Cims at outside connections is weird

Both for road and train, it's weird seeing cims at outside connection, especially as they seem to have collision mask enabled (that might have changed since I last checked) - can't they be put in `CitizenUnit` and then just transfer the CUs to car/train once it's spawned? That way there's not a pile of spawned citizens lurking at the edge of map. Might also remove need for intercity buses spawning on virtual lane (also: code for intercity buses could be applied to trains/cars at the outside connections?)

### Changing aircraft outside connections = risky

It's very easy to break the aircraft AI resulting in no more aircraft spawning. It's possible that CO will fix that as part of upcoming Airports DLC?

### Other random stuff

Not outside connection related, but it would be nice to have networks (and, optionally, terrain) render a bit better on non-purchased tiles similar to [Common Ground](https://steamcommunity.com/sharedfiles/filedetails/?id=2590879367) mod? Most of the tiles I purchased (before discovering that mod) in 81 Tiles were to make my outside routes render better. If roads/tracks just rendered properly in first place, I wouldn't need to purchase as many tiles.

**Comment by Quistar-LAB** [comment]
These are all valid points and I'll try to realize them. I'll first layout the foundation, and try to go over each one of these issues/extra functionalities

**Comment by originalfoo** [comment]
Another 81 Tiles issue I just learnt about: Intercity bus stops/stations/terminals don't work outside central 25 tile area

See comments on this workshop asset: https://steamcommunity.com/sharedfiles/filedetails/?id=2049510825

**Comment by algernon-A** [comment]
If this functionality is to be part of EML, then it makes sense to also merge Unlimited Outside Connections in as well.  Alternatively these items could be kept out of EML and instead implemented in a separate mod; I'd suggest building on the framework of Unlimited Outside Connections Revisited (all of the points above other than 'Cims at outside connections' and 'Other random stuff' are already on the to-do list for that mod anyway).

**Comment by Quistar-LAB** [comment]
I know that intercity bus has a lot of issues with 81 tiles, but is Unlimited Outside Connection broken?

**Comment by algernon-A** [comment]
Currently, the original BP version has a bug that prevents it from working in editors - BP is currently unresponsive (and my PR to fix that issue has been sitting there since 15 Jul 2021).

My own 'Unlimited Outside Connections' currently works fine - it was supposed to be a 'beta' for an update to the original, but since BP left the task unfinished, it's now continuing as a permanent replacement.   I was intending on expanding its functionality to include many of the above points, as they're all inter-related.

I'm just saying that if EML incorporates the above, then it's a very trivial addition to include unlimited outside connections as well, and no point in having two different mods do the same thing.  Either EML includes this, or I can add to Unlimited Outside Connections.  It probably comes down to what exactly needs to be patched in EML to replace 81 Tiles.

**Comment by Quistar-LAB** [comment]
Gotcha! I'll have a look at the codes and see how best to incorporate, my own version of 81 tiles is about 70% done, so.. soon I'll have to start tackling these issues

**Comment by Quistar-LAB** [comment]
As you can see, I've placed about 10 dams near the edge of the map using my 81 tiles implementation, trying to reproduce the tsunami issue... but I can't seem to reproduce it..
![Screenshot 40](https://user-images.githubusercontent.com/82305431/148109786-a92f98d5-003e-48e8-986a-c34fcbe5baa1.png)
Ok, so I'm about 90% done with 81 Tiles, and I decided to test the tsunami issue, but the problem is I can't reproduce it!! Does anyone know how to reproduce this tsunami effect when using 81 tiles?

**Comment by originalfoo** [comment]
The only thing I've seen that causes a tsunami while dam is still present is after unlocking the road on the dam (ie. clearing the `untouchable` flag) which seems to cause dam to fail. Other than that, I've never had any real problems with dams regardless of where they're placed.

**Comment by algernon-A** [comment]
I did see it happen once, but only once, and could never reproduce... however, random 'water spikes' (without dams) are a more regular occurrence for me (and no, sorry, no test cases....).  I have wondered if the specific "dam tsunamis" are actually a thing, or if it's just the more generic water spikes (with confirmation bias or similar associating them with dams).

**Comment by originalfoo** [comment]
There's probably some 'optimisation' in the vanilla water code which reduces the accuracy of calcuations / water mesh outside central 25 tile area (premise being that the user can't build there so some weirdness with water in the outer rim wouldn't be a problem in vanilla). Similar sort of thing to how the terrain and networks 'degrade' in unpurchased tiles...? That could certainly cause random flooding and 'big waves' in rivers coming from the dam due to reduced 'smoothing' of the water surface?

**Comment by originalfoo** [comment]
BTW, if you are fixing disasters stuff (Natual Disasters DLC) as part of your 81 Tiles and you come across the code that auto-moves the camera to the disaster when it happens, it would be great if that didn't happen when a tool is active. :) I've had so many annoyances with that thing, especially if I was terraforming at the time or moving stuff with Move It lol.

**Comment by algernon-A** [comment]
> BTW, if you are fixing disasters stuff (Natual Disasters DLC) as part of your 81 Tiles and you come across the code that auto-moves the camera to the disaster when it happens, it would be great if that didn't happen when a tool is active. :) I've had so many annoyances with that thing, especially if I was terraforming at the time or moving stuff with Move It lol.

Funnily enough, I'm working on a camera mod that includes that functionality!

**Comment by Quistar-LAB** [comment]
> > BTW, if you are fixing disasters stuff (Natual Disasters DLC) as part of your 81 Tiles and you come across the code that auto-moves the camera to the disaster when it happens, it would be great if that didn't happen when a tool is active. :) I've had so many annoyances with that thing, especially if I was terraforming at the time or moving stuff with Move It lol.
> 
> Funnily enough, I'm working on a camera mod that includes that functionality!

You can do it in this camera mod of yours, since Camera framework is independent of GameAreaManager, DistrictManager nor TerrainManager.

**Comment by originalfoo** [comment]
@algernon-A Is the camera mod repo available? I have a couple of feature requests (acutally dozens, but I limit to 2 here hehe) :)

1. Turn off rain when underground
    * [First Person Camera: Updated](https://steamcommunity.com/sharedfiles/filedetails/?id=650805785&searchtext=first+person+camera) mod can display road name that the vehicle is on, which means it can get the segment netinfo and check underground flag
2. When vehicle is about to go round a bend, it should be possible to pan the camera in to the turn a little which will improve the experience for viewer
    * I experimented with that manually using [First-person Auto-walking](https://steamcommunity.com/sharedfiles/filedetails/?id=659877005&searchtext=first+person+camera) mod which allows camera pan/tilt via mouse; "looking in to the turn" was a huge improvement over just looking straight ahead at scenery zooming sideways
    * @krzychu124 has a mod for visualising the pathUnit which could likely give some ideas as to how to determine curvature (vehicle speed vs. simulation accuracy will need taking in to account): https://github.com/CitiesSkylinesMods/TMPE/issues/1229#issuecomment-995387608

**Comment by algernon-A** [comment]
Camera mod repo not yet available - hopefully soon once I settle on architectural approach.  Will let you know when it is!

**Comment by originalfoo** [comment]
@Quistar-LAB I think there's a possible audio issue when lightning strikes (thunderstorm) occur outside central 25 tile area - it seems to cause audio to fail, most notably on vehicles, but eventually other audio fails too including buildings and eventually even radio (music). A lightning strike within central 25 area is not causing issues.

I have a huge cacophony of mods so it could be something like Ambient Sounds Tuner or Vehicle Effects, etc., but just flagging it up incase you spot anything while doing your version of 81 Tiles.

EDIT: I also have [Natural Diasters Overhaul](https://steamcommunity.com/sharedfiles/filedetails/?id=1801953480) mod which increases likelihood that disasters happen outside city while the city is small.

**Comment by originalfoo** [comment]
Hydro Dam tsunamis outside 25 tile area seem to be real thing. As far as I can tell it's caused when sea height changes (eg. if user has Extra Landscaping Tools mod, or Tides mod).

<details><summary>screenshot</summary>

![image](https://user-images.githubusercontent.com/1386719/150450605-3b1e85a6-6f6a-437c-a5ec-1663fe9a149e.png)

</details>

**Comment by Leo512bit** [comment]
On the note of disasters info view of where disaster may strike a is a bit buggy when you unlock enough tiles out of the 25 tile area (see [here](https://youtu.be/rtbJviCn6dQ?t=23)[This is the best demonstration of the the bug without me loading up a city right now]) the buildings are supposed to be shaded with red not multicolored and the map flickers if you press play. Also the mod page says:
>Disasters and evacuaction don't work outside of 25 tiles area (too much work, won't be fixed)

Although I remember them working fine.

Are/will those issues be fixed in this implementation of the 81 tile mod?

**Comment by CrankyAnt** [comment]
If it helps, I have a save (modded and lots of assets) where I always have water spikes flooding a part of the city in the 81-tile area after I loaded up the save. Its the part called "Elk Town" on the left-side of the map, it has a manually added canal cutting through it which causes problems after loading. 
[AutoSave.zip](https://github.com/Quistar-LAB/EManagersLib/files/8095654/AutoSave.zip)

**Comment by Monniasza** [comment]
The game does not load when all tiles have been bought.
[NewSave2.zip](https://github.com/Quistar-LAB/EManagersLib/files/8234910/NewSave2.zip)
[CompatibilityReport.txt](https://github.com/Quistar-LAB/EManagersLib/files/8234911/CompatibilityReport.txt)
[output_log.txt](https://github.com/Quistar-LAB/EManagersLib/files/8234913/output_log.txt)
